# Chapter 03 제출 답안 양식. 데이터의 첫인상 읽기

## 0. 제출 정보
- 이름: 유은송
- GitHub ID: song-03
- 작성일: 2026.09.15.
- 최종 제출 URL: https://github.com/song-03/llm-data-analysis-study/blob/main/chapter03/chapter03.ipynb

## 1. 실습 준비


In [58]:
from pathlib import Path
import sys

print(sys.executable)
print(Path.cwd())

c:\dev\llm-data-analysis-course\.venv\Scripts\python.exe
c:\dev\llm-data-analysis-course\notebooks


In [59]:
from pathlib import Path

import pandas as pd

pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 120)


def find_project_root(start: Path) -> Path:
    """현재 위치에서 위로 올라가며 data/raw 폴더가 있는 프로젝트 루트를 찾습니다."""
    start = start.resolve()
    for candidate in [start, *start.parents]:
        if (candidate / "data" / "raw").exists() and (candidate / "book").exists():
            return candidate
    raise FileNotFoundError("프로젝트 루트를 찾지 못했습니다. 노트북을 저장소 안에서 실행해 주세요.")


PROJECT_ROOT = find_project_root(Path.cwd())
DATA_DIR = PROJECT_ROOT / "data" / "raw"

print("프로젝트 루트:", PROJECT_ROOT)
print("데이터 폴더:", DATA_DIR)


프로젝트 루트: C:\dev\llm-data-analysis-course
데이터 폴더: C:\dev\llm-data-analysis-course\data\raw


## 2. 데이터 파일이 있는지 확인하기

In [60]:
expected_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",
]

file_check = pd.DataFrame({
    "file": expected_files,
    "path": [str(DATA_DIR / filename) for filename in expected_files],
    "exists": [(DATA_DIR / filename).exists() for filename in expected_files],
})

file_check


,file,path,exists
0,customers.csv,C:\dev\llm-data-analysis-course\data\raw\custo...,True
1,products.csv,C:\dev\llm-data-analysis-course\data\raw\produ...,True
2,orders.csv,C:\dev\llm-data-analysis-course\data\raw\order...,True
3,order_items.csv,C:\dev\llm-data-analysis-course\data\raw\order...,True


## 3. CSV 파일을 DataFrame으로 불러오기

In [61]:
customers = pd.read_csv(DATA_DIR / "customers.csv")
products = pd.read_csv(DATA_DIR / "products.csv")
orders = pd.read_csv(DATA_DIR / "orders.csv")
order_items = pd.read_csv(DATA_DIR / "order_items.csv")

print("customers:", type(customers))
print("products:", type(products))
print("orders:", type(orders))
print("order_items:", type(order_items))


customers: <class 'pandas.DataFrame'>
products: <class 'pandas.DataFrame'>
orders: <class 'pandas.DataFrame'>
order_items: <class 'pandas.DataFrame'>


In [62]:
datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

list(datasets.keys())


['customers', 'products', 'orders', 'order_items']

## 4. 각 파일의 역할 이해하기


| 파일 | 역할 | 먼저 확인할 것 |
| --- | --- | --- |
| `customers.csv` | 고객 정보 | 고객 수, 연령, 성별, 지역, 가입일 |
| `products.csv` | 상품 정보 | 상품 수, 카테고리, 가격 |
| `orders.csv` | 주문 정보 | 주문 수, 주문일, 결제수단, 주문상태 |
| `order_items.csv` | 주문 상세 정보 | 주문별 상품, 수량, 단가 |

처음에는 파일을 합치지 말고, 각 파일을 따로 살펴보는 것이 좋습니다.


## 5. 데이터 크기 확인하기



In [63]:
print("customers:", customers.shape)
print("products:", products.shape)
print("orders:", orders.shape)
print("order_items:", order_items.shape)


customers: (150, 6)
products: (100, 4)
orders: (300, 5)
order_items: (764, 5)


In [64]:
shape_summary = pd.DataFrame([
    {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
    }
    for name, df in datasets.items()
])

shape_summary


,dataset,rows,columns
0,customers,150,6
1,products,100,4
2,orders,300,5
3,order_items,764,5


## 6. 데이터 앞부분과 마지막 부분 보기


In [65]:
customers.head()


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-06-19
1,2,김정호,F,32,대구,2025-11-02
2,3,이경수,F,61,성남,2024-06-12
3,4,조영호,F,55,울산,2026-04-13
4,5,이예원,F,19,부산,2024-09-13


In [66]:
products.head()


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000


In [67]:
orders.head()


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-05-07,card,completed
1,2,77,2025-07-23,naver_pay,cancelled
2,3,138,2025-11-19,bank_transfer,cancelled
3,4,57,2026-01-30,kakao_pay,cancelled
4,5,125,2025-12-21,card,cancelled


In [68]:
order_items.head()


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


In [69]:
customers.tail()


,customer_id,name,gender,age,city,signup_date
145,146,김숙자,M,61,성남,2025-12-23
146,147,이정남,M,19,부산,2025-02-11
147,148,오도현,M,29,고양,2026-06-15
148,149,김정자,M,20,부산,2024-10-19
149,150,조미영,M,40,대전,2025-12-04


## 7. 컬럼명 확인하기



In [70]:
for name, df in datasets.items():
    print(f"[{name}]")
    print(list(df.columns))
    print()


[customers]
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']

[products]
['product_id', 'product_name', 'category', 'price']

[orders]
['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']

[order_items]
['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']



In [71]:
column_summary = pd.DataFrame([
    {
        "dataset": name,
        "column_count": len(df.columns),
        "column_names": ", ".join(df.columns),
    }
    for name, df in datasets.items()
])

column_summary


,dataset,column_count,column_names
0,customers,6,"customer_id, name, gender, age, city, signup_date"
1,products,4,"product_id, product_name, category, price"
2,orders,5,"order_id, customer_id, order_date, payment_met..."
3,order_items,5,"order_item_id, order_id, product_id, quantity,..."


## 8. 데이터 타입 확인하기


In [72]:
customers.info()


<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB


In [73]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    df.info()



===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):

In [74]:
dtype_summary = pd.concat(
    [df.dtypes.rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("")

dtype_summary


,customers,products,orders,order_items
customer_id,int64,,int64,
name,str,,,
gender,str,,,
age,int64,,,
city,str,,,
signup_date,str,,,
product_id,,int64,,int64
product_name,,str,,
category,,str,,
price,,int64,,


## 1. 데이터 로딩과 구조 확인
### 실행/결과
- 4개 CSV 로딩 여부: O  (`customers.csv`, `products.csv`, `orders.csv`, `order_items.csv`가 모두 존재하는 것을 확인했다. 4개 파일 모두 pandas dataframe으로 로딩 정상적으로 되었다.)
- 각 데이터 shape:(행x열 순서)	
    - customers:150x6
    - products:100x4
    - orders:300x5
    - order_items:764x5

- 주요 컬럼:
  - customers: `customer_id`, `name`, `gender`, `age`, `city`, `signup_date`
  - products: `product_id`, `product_name`, `category`, `price`
  - orders: `order_id`, `customer_id`, `order_date`, `payment_method`, `order_status`
  - order_items: `order_item_id`, `order_id`, `product_id`, `quantity`, `unit_price`

- dtypes에서 주목한 컬럼: 
    - `customers.signup_date`, `orders.order_date`: `str` 타입으로 읽힌다.      
    - `customer_id`, `product_id`, `order_id`: 2개 이상의 csv에 공통적으로 속하는 값으로, 모두 `int64` 타입으로 같은 형식으로 저장되어 있다.

### Evidence
![데이터 구조 확인](images/step01_structure.png)

### 결과 관찰
4개의 CSV 파일이 모두 존재했으며, 각각 DataFrame으로 정상적으로 불러와졌다.  
데이터 크기를 확인한 결과 customers는 150x6, products는 100x4, orders는 300x5, order_items는 764x5 (행x열)이었다. 특히 order_items의 행 수가 orders보다 많았다.

컬럼과 데이터 타입을 확인한 결과 고객 가입일인 `signup_date`와 주문일인 `order_date`는 날짜 형태의 값을 가지고 있지만 `str` 타입으로 저장되어 있었다. 반면 ID 컬럼과 `age`, `price`, `quantity`, `unit_price` 등의 컬럼은 `int64`로 확인되었다. 2개 이상의 csv에 공통적으로 속하는 `customer_id`, `product_id`, `order_id` 모두 같은 타입 `int64` 형식이었다. 


### 나의 해석과 판단
분석 전 가장 주의해서 확인해야 할 컬럼은 `signup_date`와 `order_date`라고 생각했다. 날짜를 나타내는 값으로 추정되는데, 숫자가 아닌 문자열 타입으로 현재 저장되어 있는 것으로 보인다. 따라서 기간을 설정하여 분석하기 전에 실제 날짜 타입으로 정상적으로 변환될 수 있는지 확인해야 한다.
이 외 age, price, quantity, unit_price 등 이후 숫자 크기별로 정렬하거나 계산할 때 사용해야 하는 값은 숫자형으로 지정되어 있어 추후 분석 시 편리하게 사용할 수 있을 것이라고 생각했다. 다만 `id` 항목들도 숫자형으로 지정되어 있는데, 이는 식별하기 위한 번호이기 때문에 일반적인 숫자처럼 정렬 등으로는 분석하지 않아야 한다고 생각했다. 
orders_items 행이 orders 행보다 많았는데, 이 경우 orders에 여러 아이템이 포함되어 저장되어 있을 수 있다고 생각했다. 따라서 주의하여 확인해야 한다.

### 업무·분석적 의미
데이터 구조를 확인하지 않고 바로 분석을 진행하는 경우, 지금 데이터와 같이 날짜가 문자열 상태면 문자열 상태로 분석하여 기간 분석이 잘못될 수 있다. 또한 ID 등을 일괄적으로 일반적인 숫자형 변수라고 생각해 분석하면 의미 없는 분석이 될 수 있다. 그리고 실제 컬럼명이나 데이터셋 사이 구조를 확인해야 이후 필터링을 하거나 병합을 하는 경우 잘못된 컬럼을 사용하거나 예상하지 못한 결과가 발생하는 것을 막을 수 있다. 따라서 분석 전 데이터 크기나 컬럼명, 데이터 타입, 각 컬럼이 어떤 것을 의미하는지 미리 확인하는 것이 추후 분석 오류를 줄이는 것에 도움을 줄 수 있다. 

### 한계와 추가 확인 사항
이번 실습에서는 데이터 구조와 일부 실제 값만 확인했다. 따라서 이것만으로 데이터 품질 전체가 정상이라고 판단히기는 어렵다. 결측치가 존재하는지, 중복 데이터가 존재하는지, 날짜 컬럼이 변환 가능한지, 숫자형 값의 범위 분포가 적절한지 등 추가 확인이 필요하다. 

## 9. 결측치 확인하기



In [75]:
customers.isna().sum()


customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

In [76]:
missing_summary = pd.concat(
    [df.isna().sum().rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("").astype(str)

missing_summary


,customers,products,orders,order_items
customer_id,0.0,,0.0,
name,0.0,,,
gender,0.0,,,
age,0.0,,,
city,0.0,,,
signup_date,0.0,,,
product_id,,0.0,,0.0
product_name,,0.0,,
category,,0.0,,
price,,0.0,,


In [77]:
missing_rate_summary = pd.concat(
    [(df.isna().mean() * 100).round(2).rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("")

missing_rate_summary


,customers,products,orders,order_items
customer_id,0.0,,0.0,
name,0.0,,,
gender,0.0,,,
age,0.0,,,
city,0.0,,,
signup_date,0.0,,,
product_id,,0.0,,0.0
product_name,,0.0,,
category,,0.0,,
price,,0.0,,


### 결측치 해석

| 처리 방법 | 설명 |
| --- | --- |
| 행 제외 | 결측치가 있는 행을 분석에서 제외합니다. |
| 대표값 대체 | 평균, 중앙값, 최빈값 등으로 채웁니다. |
| 별도 범주 처리 | `Unknown` 같은 범주로 표시합니다. |
| 컬럼 제외 | 분석 목적에 맞지 않는 컬럼은 사용하지 않습니다. |
| 원인 확인 | 수집 과정에서 문제가 있었는지 확인합니다. |


## 10. 중복 데이터 확인하기


In [78]:
duplicate_rows = pd.DataFrame([
    {
        "dataset": name,
        "duplicated_rows": df.duplicated().sum(),
    }
    for name, df in datasets.items()
])

duplicate_rows


,dataset,duplicated_rows
0,customers,0
1,products,0
2,orders,0
3,order_items,0


In [79]:
id_duplicate_checks = pd.DataFrame([
    {
        "check": "customers.customer_id",
        "duplicated_count": customers["customer_id"].duplicated().sum(),
        "interpretation": "0이어야 고객 ID가 유일합니다.",
    },
    {
        "check": "products.product_id",
        "duplicated_count": products["product_id"].duplicated().sum(),
        "interpretation": "0이어야 상품 ID가 유일합니다.",
    },
    {
        "check": "orders.order_id",
        "duplicated_count": orders["order_id"].duplicated().sum(),
        "interpretation": "0이어야 주문 ID가 유일합니다.",
    },
    {
        "check": "order_items.order_id",
        "duplicated_count": order_items["order_id"].duplicated().sum(),
        "interpretation": "한 주문에 여러 상품이 있으면 0보다 클 수 있습니다.",
    },
])

id_duplicate_checks


,check,duplicated_count,interpretation
0,customers.customer_id,0,0이어야 고객 ID가 유일합니다.
1,products.product_id,0,0이어야 상품 ID가 유일합니다.
2,orders.order_id,0,0이어야 주문 ID가 유일합니다.
3,order_items.order_id,464,한 주문에 여러 상품이 있으면 0보다 클 수 있습니다.


In [80]:
order_items["order_item_id"].duplicated().sum()


np.int64(0)

In [81]:
for name, df in datasets.items():
    print(name, df.duplicated().sum())

customers 0
products 0
orders 0
order_items 0


## 2. 결측·중복·키 품질 
- 주요 ID 결측: `customers.csv`, `products.csv`, `orders.csv`, `order_items.csv` 모두에서 0건
- 주요 ID 중복: `customers.csv`, `products.csv`, `orders.csv`, `order_items.csv` 모두에서 0건
- 전체 행 중복: `customers.csv`, `products.csv`, `orders.csv`, `order_items.csv` 모두에서 0건

![결측 중복 점검](images/step02_quality.png)

### 결과 관찰
4개 CSV의 결측치를 확인한 결과 모든 컬럼에서 결측치 0건, 주요 ID에서 결측치 0건이었다. 즉 결측치가 발견되지 않았다. 중복 여부 경우에도 주요 ID 중복여부를 확인한 결과 모두 중복 0건, 전체 행 중복 역시 모두 0건이었다. 
주요 ID의 중복 여부를 확인한 결과  `customer_id`, `product_id`, `order_id`, `order_item_id` 모두 중복 0건이었다. 전체 행 중복도 모두 0건이었다. 단 `order_items.order_id`에서는 중복이 464건이었다. 

### 나의 해석과 판단
먼저 4개 csv 데이터에서 주요 ID 결측, 주요 ID 중복, 전체 행 중복 모두 나타나지 않았다. 따라서 기본적인 식별을 위해서는 양호한 상태라고 판단했다. `order_items.order_id` 에서는 464건 중복이 나타났으나 잘못된 것이라고 판단하지는 않았다. 한 주문에 포함된 여러 상품을 각각 행으로 기록하는 데이터니까, 하나의 주문에 여러 상품이 포함되어 있는 경우에는 반복될 수 있기 때문이다.
중복 문제가 발견되었을 경우 단순히 중복 건수가 많은 항목부터 삭제하는 게 아니라, 먼저 중복되어서는 안되는 고유한 ID인지, 다른 테이블과 연결하기 위한 키인지 등을 확인해야 한다고 생각한다.
전체 행 중복이 발견된 경우에도 바로 삭제하기보다는 실제 중복 입력인지, 동일한 거래나 이벤트가 여러 번 발생한 정상 기록인지 확인할 필요가 있다. 따라서 결측이나 중복이 발견되더라도 먼저 컬럼의 역할과 데이터 생성 방식을 확인한 뒤 처리 방법을 결정하는 것이 적절하다고 판단했다.
만약 고유해야 하는 주요 ID에서 중복이 발견된다면, 같은 대상을 여러 번 작성한 것인지, 혹은 ID가 잘못된 것인지 등 원인을 파악한 후에, 처리 여부를 결정해야 한다. 

### 업무·분석적 의미
결측치가 있는 경우 필요한 정보가 빠져있어 분석 대상이 누락되거나 집계 결과가 실제 데이터 전체를 반영하지 못할 수 있다. 특히 주요 ID에 문제가 있으면 다른 테이블과 연결하기 어려울 수 있어 반드시 확인이 필요하다.
또한 중복 중에서도 고유해야 하는 주요 ID에 중복이 있으는 경우 데이터를 병합하거나 집계하는 과정에서 똑같은 값이 여러번 계산되어 잘못 분석될 수 있다. 이때 중복이 어떤 유형인지도 파악해야 하는데, 반복될 수 있는 연결 ID를 중복이라는 이유로 삭제하는 경우 한 주문에 포함된 여러 상품 정보가 사라져 데이터가 변형될 수 있다. 
따라서 오류 방지 및 적절한 분석을 위해서 먼저 컬럼 역할을 파악하고, 중복을 판단하는 것이 필요하다.

### 한계와 추가 확인 사항
현재 결측치, 중복치 검사를 진행해 데이터 품질을 확인하기는 했으나, 이는 일부일뿐 이것으로 데이터 전체가 정상이라고 판단하기는 어렵다. 또한 `order_items.order_id`의 반복이 실제 내가 추측한 이유가 맞는지, 실제 주문과 정상적으로 연결되는지 검증해야 한다. 

## 11. 숫자형 컬럼 기본 통계 확인하기


In [82]:
customers.describe()


,customer_id,age
count,150.000000,150.000000
mean,75.500000,42.086667
std,43.445368,15.613166
min,1.000000,19.000000
25%,38.250000,29.000000
50%,75.500000,40.000000
75%,112.750000,57.000000
max,150.000000,69.000000


In [83]:
products[["price"]].describe()


,price
count,100.000000
mean,110040.000000
std,56433.910574
min,5000.000000
25%,65750.000000
50%,112000.000000
75%,161000.000000
max,200000.000000


In [84]:
order_items[["quantity", "unit_price"]].describe()


,quantity,unit_price
count,764.000000,764.000000
mean,3.053665,108561.518325
std,1.410873,56996.770604
min,1.000000,5000.000000
25%,2.000000,62000.000000
50%,3.000000,111000.000000
75%,4.000000,161250.000000
max,5.000000,200000.000000


In [85]:
price_text = pd.Series(["10,000", "25,500", "3000", "확인필요"])
price_number = pd.to_numeric(
    price_text.str.replace(",", "", regex=False),
    errors="coerce",
)

pd.DataFrame({
    "original": price_text,
    "converted": price_number,
})


,original,converted
0,"10,000",10000.0
1,"25,500",25500.0
2,3000,3000.0
3,확인필요,NaN


## 12. 범주형 컬럼 고유값 확인하기


In [86]:
customers["city"].value_counts().head(10)


city
성남    21
광주    17
부산    16
대구    15
서울    15
울산    14
인천    14
대전    14
수원    13
고양    11
Name: count, dtype: int64

In [87]:
products["category"].value_counts()


category
스포츠     19
전자기기    17
생활용품    16
뷰티      16
도서      14
패션      11
식품       7
Name: count, dtype: int64

In [88]:
orders["order_status"].value_counts()


order_status
completed    184
cancelled     64
refunded      52
Name: count, dtype: int64

In [89]:
categorical_summary = pd.DataFrame([
    {"dataset": "customers", "column": "city", "unique_count": customers["city"].nunique()},
    {"dataset": "customers", "column": "gender", "unique_count": customers["gender"].nunique()},
    {"dataset": "products", "column": "category", "unique_count": products["category"].nunique()},
    {"dataset": "orders", "column": "payment_method", "unique_count": orders["payment_method"].nunique()},
    {"dataset": "orders", "column": "order_status", "unique_count": orders["order_status"].nunique()},
])

categorical_summary


,dataset,column,unique_count
0,customers,city,10
1,customers,gender,2
2,products,category,7
3,orders,payment_method,4
4,orders,order_status,3


## 13. 날짜 컬럼 확인하기


In [90]:
orders["order_date"].head()


0    2026-05-07
1    2025-07-23
2    2025-11-19
3    2026-01-30
4    2025-12-21
Name: order_date, dtype: str

In [91]:
print("변환 전 타입:", orders["order_date"].dtype)

orders["order_date"] = pd.to_datetime(orders["order_date"], errors="coerce")

print("변환 후 타입:", orders["order_date"].dtype)
print("날짜 변환 실패 건수:", orders["order_date"].isna().sum())
print("가장 빠른 주문일:", orders["order_date"].min())
print("가장 최근 주문일:", orders["order_date"].max())


변환 전 타입: str
변환 후 타입: datetime64[us]
날짜 변환 실패 건수: 0
가장 빠른 주문일: 2025-07-09 00:00:00
가장 최근 주문일: 2026-07-08 00:00:00


In [92]:
###가입일에 대해서도 확인
signup_check = pd.to_datetime(
    customers["signup_date"],
    errors="coerce",
)

print("가입일 변환 실패:", signup_check.isna().sum())

가입일 변환 실패: 0


In [93]:
orders.assign(
    order_year=orders["order_date"].dt.year,
    order_month=orders["order_date"].dt.month,
    order_day_name=orders["order_date"].dt.day_name(),
).head()


,order_id,customer_id,order_date,payment_method,order_status,order_year,order_month,order_day_name
0,1,123,2026-05-07,card,completed,2026,5,Thursday
1,2,77,2025-07-23,naver_pay,cancelled,2025,7,Wednesday
2,3,138,2025-11-19,bank_transfer,cancelled,2025,11,Wednesday
3,4,57,2026-01-30,kakao_pay,cancelled,2026,1,Friday
4,5,125,2025-12-21,card,cancelled,2025,12,Sunday


## 3. 숫자형·범주형·날짜 점검
- 숫자형 범위에서 주목한 값
  - `age`: 평균 약 42세, 최소 19세, 최대 69세
  - `price`: 평균 110,040원, 최소 5,000원, 최대 200,000원
  - `quantity`: 평균 약 3개, 최소 1개, 최대 5개
  - `unit_price`: 평균 약 108,562원, 최소 5,000원, 최대 200,000원

- 범주형 빈도에서 주목한 값: 
  - `city`: 총 10개 지역, 성남이 21명으로 가장 많고, 고양이 11명으로 가장 적음
  - `category`: 총 7개 범주, 스포츠 19개가 가장 많고, 식품 7개가 가장 적음
  - `order_status`: completed 184건, cancelled 64건, refunded 52건
  - 범주형 고유값 수는 `gender` 2개, `payment_method` 4개로 확인되었다.

- 날짜 변환 실패 건수: 0건   (주문일, 가입일 모두)
- 날짜 범위: 2025-07-09 ~ 2026-07-08

![기본 분포와 날짜 확인](images/step03_distribution.png)

### 결과 관찰
Describe() 함수를 이용해 각 csv의 숫자형 컬럼의 개수, 평균, 표준편차, 최솟값, 사분위수, 최댓값을 확인한 결과  customer age의 경우 평균 약 42세, 최소 19세, 최대 69세로 나타났다. Price의 경우 퍙군 110040원, 최소 5천원, 최대 200000원이었다. order_items의 quantity 경우 평균 약 3개, 최소 1개, 최대 5개였다. unit_price 경우 평균 약 108562원, 최소 5천원, 최대 200000원이었다. 
범주형 데이터를 확인한 결과 `city`는 10개의 값으로 구성되어 있으며 성남이 21명으로 가장 많고 고양이 11명으로 가장 적었다. 상품 `category`는 총 7개였으며 스포츠가 19개로 가장 많고 식품이 7개로 가장 적었다. 주문 상태는 completed 184건, cancelled 64건, refunded 52건으로 completed가 가장 많았다. gender는 2개의 값으로 구성되어 있었다. 
`orders.order_date`의 경우 기존에는 `str` 타입으로 저장되어 있었는데, `pd.to_datetime()`을 이용하여 날짜 타입으로 변환한 결과 변환 실패 0건이었다. 가장 빠른 주문일은 2025-07-09, 가장 최근 주문일은 2026-07-08 였다. order year, month, 요일, payment method, order status 등까지 확인 가능했다. 


### 나의 해석과 판단

현재 확인된 결과에서 최솟값 및 최댓값만 보았을 때 비정상적으로 보이는 값은 없었다. 모두 양수값이었으며, 최대 최소값의 경우 정수 형태로 나타났다. 다만 값이 범위 안에 있단 사실만으로 정확한 데이터라고 단정하기는 어렵다. 특히 평균 가격이 10만원 이상으로 높아 이 평균값이 실제로 item의 가격을 전체적으로 반영하는 값인지, 혹은 최댓값 등이 영향을 크게 준 것인지 확인해야 한다. outlier 값에 의해 나타난 평균인 경우 중앙값 역시 봐야 한다. 이 데이터 경우 중앙값이 unit_price 기준 111000원, 평균 108561.5원이므로 비교적 전체 데이터를 잘 반영한다고 볼 수 있다. 다만 price와 unit_price의 최소, 최댓값은 같으나 총 count나 mean이 약간 다르므로 그 차이가 어디에서 기인하는 것인지 확인해야 한다.
또한 숫자로 보이는 컬럼이라도 실제 역할을 구분해서 해석해야 한다고 생각했다. `customer_id`, `product_id`, `order_id`, `order_item_id`는 `int64`로 저장되어 있으나 식별을 위한 값이기 때문에 평균, 크기 등을 분석하는 것은 의미가 없닥다고 생각했다. 또한 숫자가 문자열로 저장되는 경우도 있을 수 있으므로 데이터 타입을 먼저 확인해야 한다. 다만 이번 실습에서는 이전 단계에서 해당 열의 타입을 모두 확인했고, 날짜 제외 다른 숫자형으로 되어있어야 할 항목은 int64 형태임을 확인했다.
또한 범주형 데이터 (order_status) 의 경우 빈도를 살펴봐야 하는데, 빈도가 특정값이 높거나 낮다는 이유만으로 이상값이라고 생각해서는 안된다. completed된 주문이 많은데, 이것이 정상적인 결과인지 혹은 데이터 수집 과정에서 편향되게 수집된 것인지 현재 결과로 파악이 어려우므로 추가적인 조사가 필요하다. 어떻게 이 데이터를 받았는지, 어떤 필터나 제약 조건 하에서 수집한 결과인지 살펴봐야 한다. 

### 업무·분석적 의미
숫자형 데이터의 범위와 분포를 확인해야 하는데, 이를 확인하지 않고 분석하는 경우 입력 오류나 비정상적인 값이 통계값(평균 등)에 영향을 줘 잘못된 해석을 유도할 수 있다. 따라서 분석 전에 최댓값, 최소값 등을 확인하고 범위를 확인해 정상적인 값인지 검토해야 한다. 범주형 데이터 경우에도 빈도를 미리 확인해 특정 값에 데이터가 지나치게 집중되어 있지는 않은지, 예상하지 못한 범주도 존재하는지 확인해야 한다. 이를 확인하지 않고 비교를 진행하면 표본 수 차이를 고려하지 않은 채 결과를 해석하게 될 수도 있다. 또한 날짜 컬럼 경우 미리 문자열 상태에서 날짜 타입으로 변환해야 기간별 분석이 가능하다. 

### 한계와 추가 확인 사항
현재 숫자형 데이터에서 본 기본적인 사항으로는 실제로 개별 값이 실제 업무적으로 정상적인 값인지 판단하기 어렵다. 특히 상품별로 적정가가 존재하는데, 이 값에 해당하는 값인지는 실제 데이터 값을 확인해야 한다. 또한 한 주문에서 가능한 최대 수량이나 최대값 등 기준이 있는 경우 이 역시 확인해야 한다. 
또한 범주형 데이터에서 분포가 두드러지게 나타나는 경우 그것이 데이터 수집 과정에서 발생한 편향의 문제인지, 실제로 나타나는 분포인지 확인해야 한다. 
price와 unit_price의 최소, 최댓값은 같으나 총 count나 mean이 약간 다르므로 그 차이가 어디에서 기인하는 것인지 확인해야 한다.

## 14. 여러 파일의 관계 확인하기


| 연결 관계 | 의미 |
| --- | --- |
| `customers.customer_id` ↔ `orders.customer_id` | 어떤 고객이 주문했는지 연결합니다. |
| `orders.order_id` ↔ `order_items.order_id` | 주문과 주문 상세를 연결합니다. |
| `products.product_id` ↔ `order_items.product_id` | 주문 상세와 상품 정보를 연결합니다. |


In [94]:
invalid_customers = orders[~orders["customer_id"].isin(customers["customer_id"])]
invalid_orders = order_items[~order_items["order_id"].isin(orders["order_id"])]
invalid_products = order_items[~order_items["product_id"].isin(products["product_id"])]

relationship_check = pd.DataFrame([
    {
        "relationship": "orders.customer_id -> customers.customer_id",
        "invalid_rows": len(invalid_customers),
    },
    {
        "relationship": "order_items.order_id -> orders.order_id",
        "invalid_rows": len(invalid_orders),
    },
    {
        "relationship": "order_items.product_id -> products.product_id",
        "invalid_rows": len(invalid_products),
    },
])

relationship_check


,relationship,invalid_rows
0,orders.customer_id -> customers.customer_id,0
1,order_items.order_id -> orders.order_id,0
2,order_items.product_id -> products.product_id,0


## 15. 간단한 병합으로 관계 확인하기



In [95]:
order_items_with_products = order_items.merge(
    products,
    on="product_id",
    how="left",
)

order_items_with_products["line_amount"] = (
    order_items_with_products["quantity"] * order_items_with_products["unit_price"]
)

order_items_with_products.head()


,order_item_id,order_id,product_id,quantity,unit_price,product_name,category,price,line_amount
0,1,1,100,3,102000,도서 상품 100,도서,102000,306000
1,2,1,87,5,25000,도서 상품 087,도서,25000,125000
2,3,1,7,3,142000,도서 상품 007,도서,142000,426000
3,4,1,9,3,193000,스포츠 상품 009,스포츠,193000,579000
4,5,2,72,4,189000,뷰티 상품 072,뷰티,189000,756000


In [96]:
category_sales = (
    order_items_with_products
    .groupby("category", as_index=False)["line_amount"]
    .sum()
    .sort_values("line_amount", ascending=False)
)

category_sales


,category,line_amount
3,스포츠,50174000
1,뷰티,47551000
5,전자기기,41003000
2,생활용품,34839000
4,식품,33597000
0,도서,24645000
6,패션,23801000


## 16. 반복 점검을 함수로 정리하기


In [97]:
def check_data_overview(name: str, df: pd.DataFrame) -> None:
    print(f"===== {name} =====")
    print("shape:", df.shape)
    print("\ncolumns:")
    print(list(df.columns))
    print("\ndtypes:")
    print(df.dtypes)
    print("\nmissing values:")
    print(df.isna().sum())
    print("\nduplicated rows:", df.duplicated().sum())


check_data_overview("customers", customers)


===== customers =====
shape: (150, 6)

columns:
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']

dtypes:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object

missing values:
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

duplicated rows: 0


In [98]:
for name, df in datasets.items():
    check_data_overview(name, df)
    print()


===== customers =====
shape: (150, 6)

columns:
['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']

dtypes:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object

missing values:
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

duplicated rows: 0

===== products =====
shape: (100, 4)

columns:
['product_id', 'product_name', 'category', 'price']

dtypes:
product_id      int64
product_name      str
category          str
price           int64
dtype: object

missing values:
product_id      0
product_name    0
category        0
price           0
dtype: int64

duplicated rows: 0

===== orders =====
shape: (300, 5)

columns:
['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']

dtypes:
order_id                   int64
customer_id                int64
order_date        datetime64[us]
payment_method          

In [99]:
def check_data_overview(name, df, key_column=None):
    print(f"===== {name} =====")
    print("shape:", df.shape)
    print("columns:", list(df.columns))
    print("dtypes:")
    print(df.dtypes)
    print("missing:")
    print(df.isna().sum())
    print("duplicated rows:", df.duplicated().sum())

    if key_column is not None:
        print("key missing:", df[key_column].isna().sum())
        print("key duplicated:", df[key_column].duplicated().sum())

key_columns = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}

for name, df in datasets.items():
    check_data_overview(
        name,
        df,
        key_column=key_columns[name]
    )
    print()

===== customers =====
shape: (150, 6)
columns: ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
dtypes:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object
missing:
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64
duplicated rows: 0
key missing: 0
key duplicated: 0

===== products =====
shape: (100, 4)
columns: ['product_id', 'product_name', 'category', 'price']
dtypes:
product_id      int64
product_name      str
category          str
price           int64
dtype: object
missing:
product_id      0
product_name    0
category        0
price           0
dtype: int64
duplicated rows: 0
key missing: 0
key duplicated: 0

===== orders =====
shape: (300, 5)
columns: ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
dtypes:
order_id                   int64
customer_id                int64
order_date     

## 4. CSV 간 키 관계 검증 
- 없는 `customer_id`: 없음 (`orders.customer_id → customers.customer_id`)
- 없는 `order_id`: 없음 (`order_items.order_id → orders.order_id`)
- 없는 `product_id`: 없음 (`order_items.product_id → products.product_id`)

![PK FK 관계 검증](images/step04_relationship.png)

### 결과 관찰
csv 사이 키 관계를 확인한 결과, `orders.customer_id → customers.customer_id`, `order_items.order_id → orders.order_id`, `order_items.product_id → products.product_id` 에 존재하지 않는 값 모두 0이었다. 
키 관계를 확인한 후 `order_items`와 `products`를 `product_id` 기준으로 병합한 결과 상품명, 카테고리 등 상품 정보가 주문 상세 데이터에 연결되었다. 이후 quantity * unit_price를 통해 각 행의 line_amount를 계산하고 카테고리 별로 집계할 수 있었다.

### 나의 해석과 판단
csv 사이 키 관계를 확인한 결과, `orders.customer_id → customers.customer_id`, `order_items.order_id → orders.order_id`, `order_items.product_id → products.product_id` 에 존재하지 않는 값 모두 0이었으므로 현재 데이터에서 주문 데이터의 고객 ID, 주문 상세 데이터의 주문 ID, 상품 ID가 각각 연결 대상인 부모 테이블에 모두 존재하는 것이라고 해석할 수 있었다. 
현재 확인한 범위에서는 키 연결이 적절히 되어있으며 csv 사이 기본적인 연결 관계가 있었다. 따라서 이를 통해 데이터 병합 시 이 ID들을 연결 기준으로 사용할 수 있을 것이라고 생각했다.
다만 연결되지 않은 ID가 있다고 해서 해당 행을 바로 삭제해서는 안 된다고 생각했다. 부모 테이블 데이터가 누락되었거나, 두 파일의 ID의 데이터 타입이 다르거나, 문자열의 앞 뒤 공백이 다른 등 형식의 작은 차이가 있을 수 있다. 혹은 파일의 수집 시점이 달라서 한쪽에만 데이터가 존재하는 경우가 있을 수도 있다. 따라서 키 관계에 문제가 있는 경우, 단순히 문제가 있는 행을 제거하기보다는 어떤 테이블에서 문제가 발생했는지 확인하고, 그 원인을 확인한 뒤에 처리 방법을 결정해야 한다고 생각했다. 

### 업무·분석적 의미
csv 사이 키 관계를 확인하는 것은 추후 데이터 병합 과정 등에서 필요한 과정이다. 정상적으로 키 관계가 연결되어 있지 않은 경우 병합하는 과정에서 주문이나 상품 정보가 일부 누락될 수도 있고, 키가 중복되어있거나 잘못 연결된 경우에는 하나 행이 여러 행으로 계산되어 실제보다 크게 반영될 수도 있다. 따라서 csv 분석 전 사용할 ID가 서로 잘 대응하는지 확인하는 것이 중요하다.

관계에서 연결되지 않는 ID가 없었기 때문에 상품 정보를 병합하고 `line_amount`를 계산하는 과정까지 정상적으로 진행할 수 있었다. 다만 이 결과는 본격적인 EDA 결과라기보다 데이터가 실제로 연결 가능한지를 확인하기 위한 기본 점검으로 보는 것이 적절하다.

### 한계와 추가 확인 사항
이번 키 관계를 확인하는 과정을 통해 세 관계 모두 정상적으로 연결되어있다고 판단할 수 있었으나, 이것만으로 데이터의 전체적인 정확성이 보장된다고 보기에는 어렵다. 또한 ID가 정상적으로 연결된다고 해도, 실제 고객 - 주문 - 상품 등이 바르게 대응되어 있는지는 확인할 수 없다. 이번 실습은 csv 사이 연결 구조가 제대로 유지되고 있는지만 파악한 것이라고 볼 수 있다. 따라서 구체적인 분석 기준이나 값 자체가 업무적으로 정확한지는 추후 확인해야 한다.

## 17. LLM에게 데이터 구조를 설명시키는 법

In [100]:
llm_dataset_summary = shape_summary.merge(column_summary, on="dataset")
llm_dataset_summary


,dataset,rows,columns,column_count,column_names
0,customers,150,6,6,"customer_id, name, gender, age, city, signup_date"
1,products,100,4,4,"product_id, product_name, category, price"
2,orders,300,5,5,"order_id, customer_id, order_date, payment_met..."
3,order_items,764,5,5,"order_item_id, order_id, product_id, quantity,..."


In [101]:
for _, row in llm_dataset_summary.iterrows():
    print(f"- {row['dataset']}: {row['rows']}행 {row['columns']}열")
    print(f"  컬럼: {row['column_names']}")


- customers: 150행 6열
  컬럼: customer_id, name, gender, age, city, signup_date
- products: 100행 4열
  컬럼: product_id, product_name, category, price
- orders: 300행 5열
  컬럼: order_id, customer_id, order_date, payment_method, order_status
- order_items: 764행 5열
  컬럼: order_item_id, order_id, product_id, quantity, unit_price


In [102]:
validation_check = pd.DataFrame([
    {
        "question": "customers에 age 컬럼이 있는가?",
        "result": "age" in customers.columns,
    },
    {
        "question": "age 결측치 개수는?",
        "result": customers["age"].isna().sum() if "age" in customers.columns else "컬럼 없음",
    },
    {
        "question": "orders.customer_id가 customers.customer_id와 연결되는가?",
        "result": len(invalid_customers) == 0,
    },
    {
        "question": "매출 계산에 필요한 quantity와 unit_price가 있는가?",
        "result": {"quantity", "unit_price"}.issubset(order_items.columns),
    },
])

validation_check


,question,result
0,customers에 age 컬럼이 있는가?,True
1,age 결측치 개수는?,0
2,orders.customer_id가 customers.customer_id와 연결되는가?,True
3,매출 계산에 필요한 quantity와 unit_price가 있는가?,True


## 5. LLM 구조 설명 검증
- LLM에 제공한 Safe Context: 
    온라인 쇼핑몰 데이터를 분석할 것임. 그 전 CSV 파일 구조를 확인하려고 함.
    데이터셋 요약:
        dataset  rows  columns                                                    column_names
    customers   150        6               customer_id, name, gender, age, city, signup_date
    products   100        4                       product_id, product_name, category, price
        orders   300        5 order_id, customer_id, order_date, payment_method, order_status
    order_items   764        5       order_item_id, order_id, product_id, quantity, unit_price

    파일의 관계:
        - customers.customer_id -> orders.customer_id
        - orders.order_id -> order_items.order_id
        - products.product_id -> order_items.product_id

    해줬으면 하는 것:
    1. 각 파일이 어떤 역할을 하는지 설명할 것.
    2. 분석 전 확인하면 좋을 항목 체크리스트를 작성할 것.
    3. 실제 데이터 확인 없이 단정지은 내용과 추가적인 확인이 필요한 내용을 구분할 것

- LLM이 제안한 추가 점검:
    - 키와 관계: 기본키 중복과 테이블 간 외래키 연결이 정상적인지 확인
    - 결측치와 중복: NULL, 빈값, 중복 행이 있는지 확인
    - 데이터 타입과 값 범위: 숫자·날짜 형식이 올바르고 비정상 값이 없는지 확인
    - 범주형 값: 범주값의 오탈자, 공백, 대소문자 차이로 값이 나뉘지 않았는지 확인
    - 금액 관련: 상품 가격과 실제 판매 단가 차이, 할인·배송비·세금 반영 여부 확인
    - 시간 관계: 가입일과 주문일 순서, 분석 기간 및 날짜 누락 여부 확인
    - 분석 단위: 고객 수, 주문 수, 판매량, 매출의 계산 기준과 취소·환불 처리 기준 확인

- 실제 데이터에서 확인한 항목:
    - 키와 관계: 기본키 중복과 테이블 간 외래키 연결이 정상적인지 확인했다. (0건으로 모두 연결되었다)
    - 결측치와 중복: 주요 ID에서 결측과 중복이 모두 0건임을 확인했고, 전체 행 중복 역시 0건임을 확인했다. order_items.order_id 는 중복이 있었으나, 이는 한 주문에 여러 상품이 포함될 수 있어 반복될 수 있는 값임을 확인했다. 
    - 데이터 타입과 값 범위: 범주형 컬럼 값 종류, 빈도를 확인했고 숫자형 컬럼의 범위를 확인했다. 날짜 경우 처음 문자열인 것을 날짜 타입으로 변환 가능한지 확인하였고, 변환 성공했음을 확인했다. 
    - 시간 관계: 주문 날짜의 범위가 2025-07-09부터 2026-07-08까지임을 확인했다. 다만 LLM이 제안한 가입일과 주문일의 선후 관계까지는 아직 확인하지 않았다.



- 채택/수정/보류한 내용:
    - 채택: 각 값 범위 및 빈도 확인, CSV 간 키 관계 확인, 중복, 결측 확인
    - 수정: LLM이 제안안 중복 경우 뭉뚱그려 '중복'이라고만 판단하였는데, 이를 고유해야 하는 ID에 한해서만 중복을 확인하는 것으로 수정한다.  
    - 보류: 
        - 할인, 배송비, 세금 등이 반영되어 있는지 여부 확인
        - 상품 가격과 실제 판매 단가 차이 확인
        - 취소, 환불 처리 기준 및 매출 계산 기준, 판매량 기준 확인
        이 항목들은 현재까지 진행한 점검만으로는 충분히 확인할 수 없었던 정보들이기 때문에 보류하였다. 

![LLM 구조 검토](images/step05_llm.png)

### 나의 해석과 판단
LLM이 제안한 것 중 가장 유용했던 것은 앞서 `보류`에 작성한 항목들이다. 특히 할인이 들어간 상품 경우 분석 결과에 영향을 미칠 수 있기 때문에 해당 분석 기간 동안 할인이 진행되었는지 등을 분석 시 꼭 고려해야 한다고 생각한다. 따라서 추후 가능하다면 이에 대한 것 역시 추가로 확인해야 한다. LLM이 제안하기 전 생각하지 못했던 항목이었기 때문에 유용하다고 생각했으나 정보가 없어 확인하지는 못했다. 또한 LLM이 제안한 결측치, 중복, 데이터 타입, 숫자형/범주형 데이터 확인, 키 연결 확인 등의 항목은 실습에서도 진행했을 만큼 데이터 분석에서 기본적인 사항이었기 때문에 이를 되짚어 주는 면에서 유용하다고 생각했다.
반면 LLM의 제안을 그대로 받아들이는 것 역시 주의해야 한다고 생각했다. 예를 들어 중복에 대한 내용 경우, 중복되어서는 안되는 컬럼과 중복이 나타날 수 있는 컬럼을 구분해야 하는데 LLM은 그를 구분하지 않았다. 만약 LLM이 중복값을 제거하라고 제안했으면 유의미한 데이터를 지웠을 수도 있으므로 제안을 무조건적으로 수용해서는 안 된다. 


### 한계와 추가 확인 사항
LLM 제안 중 일부는 현재 가진 정보만으로는 확인하기 어려웠다. 예를 들어 할인, 배송비, 세금 반영 여부, 가입일과 주문일의 선후 관계 등은 추가적인 정보가 필요하다. 또한 데이터만으로는 취소나 환불 주문을 매출이나 주문수에 포함할 것인지 결정할 수 없다. 따라서 분석 기준을 어떻게 잡을 것인지 추가적으로 검토가 필요하다. 또한 LLM의 경우 분석 시 고려해야 할 정보 (중복의 유형 등)를 제외하고 언급할 수 있으므로 참고자료로만 사용하고 사람이 판단해야 한다.

## 19. 이번 장 점검 체크리스트

| 점검 항목 | 확인 |
| --- | --- |
| 필요한 CSV 파일이 모두 존재하는가? | [x] |
| 각 데이터셋의 행과 열 개수를 확인했는가? | [x] |
| 컬럼명이 예상과 일치하는가? | [x] |
| 날짜 컬럼의 데이터 타입을 확인했는가? | [x] |
| 숫자 컬럼이 실제 숫자형으로 저장되어 있는가? | [x] |
| 결측치가 있는 컬럼을 확인했는가? | [x] |
| 중복 데이터가 있는지 확인했는가? | [x] |
| 주요 ID 컬럼의 중복 여부를 확인했는가? | [x] |
| 여러 파일을 연결할 키 컬럼을 확인했는가? | [x] |
| 파일 간 키 관계가 실제로 연결 가능한지 확인했는가? | [x] |
| LLM에 원본 데이터 대신 구조 요약만 입력했는가? | [x] |
| LLM이 제안한 설명을 실제 데이터와 비교해 검증했는가? | [x] |


## 20. 실습 과제

아래 과제를 직접 해결해 보세요.

1. 4개 CSV 파일의 행과 열 개수를 하나의 표로 정리하세요.
2. 각 파일의 결측치 개수와 결측치 비율을 확인하세요.
3. `orders.order_date`를 날짜 타입으로 변환하고 데이터 기간을 확인하세요.
4. `order_items`와 `products`를 병합해 카테고리별 주문 금액을 계산하세요.
5. LLM에게 데이터 구조 요약을 전달하는 프롬프트를 직접 작성하세요.
6. LLM이 만든 분석 아이디어가 실제 컬럼과 키 관계에 맞는지 검증하세요.


In [103]:
'''1. 4개 CSV 파일의 행과 열 개수를 하나의 표로 정리하세요.'''
shape_summary = pd.DataFrame([
    {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
    }
    for name, df in datasets.items()
])

shape_summary

,dataset,rows,columns
0,customers,150,6
1,products,100,4
2,orders,300,5
3,order_items,764,5


In [104]:
'''2. 각 파일의 결측치 개수 확인하세요.'''
missing_summary = pd.concat(
    [df.isna().sum().rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("").astype(str)

missing_summary

,customers,products,orders,order_items
customer_id,0.0,,0.0,
name,0.0,,,
gender,0.0,,,
age,0.0,,,
city,0.0,,,
signup_date,0.0,,,
product_id,,0.0,,0.0
product_name,,0.0,,
category,,0.0,,
price,,0.0,,


In [105]:
'''2. 각 파일의 결측치 비율을 확인하세요.'''
missing_rate_summary = pd.concat(
    [(df.isna().mean() * 100).round(2).rename(name) for name, df in datasets.items()],
    axis=1,
).fillna("")

missing_rate_summary

,customers,products,orders,order_items
customer_id,0.0,,0.0,
name,0.0,,,
gender,0.0,,,
age,0.0,,,
city,0.0,,,
signup_date,0.0,,,
product_id,,0.0,,0.0
product_name,,0.0,,
category,,0.0,,
price,,0.0,,


In [106]:
'''3. `orders.order_date`를 날짜 타입으로 변환하고 데이터 기간을 확인하세요.'''
orders["order_date"] = pd.to_datetime(orders["order_date"], errors="coerce")
orders.assign(
    order_year=orders["order_date"].dt.year,
    order_month=orders["order_date"].dt.month,
    order_day_name=orders["order_date"].dt.day_name(),
).head()

,order_id,customer_id,order_date,payment_method,order_status,order_year,order_month,order_day_name
0,1,123,2026-05-07,card,completed,2026,5,Thursday
1,2,77,2025-07-23,naver_pay,cancelled,2025,7,Wednesday
2,3,138,2025-11-19,bank_transfer,cancelled,2025,11,Wednesday
3,4,57,2026-01-30,kakao_pay,cancelled,2026,1,Friday
4,5,125,2025-12-21,card,cancelled,2025,12,Sunday


In [107]:
'''4. `order_items`와 `products`를 병합해 카테고리별 주문 금액을 계산하세요.'''
order_items_with_products = order_items.merge(
    products,
    on="product_id",
    how="left",
)

order_items_with_products["line_amount"] = (
    order_items_with_products["quantity"] * order_items_with_products["unit_price"]
)
category_sales = (
    order_items_with_products
    .groupby("category", as_index=False)["line_amount"]
    .sum()
    .sort_values("line_amount", ascending=False)
)

category_sales

,category,line_amount
3,스포츠,50174000
1,뷰티,47551000
5,전자기기,41003000
2,생활용품,34839000
4,식품,33597000
0,도서,24645000
6,패션,23801000


5. LLM에게 데이터 구조 요약을 전달하는 프롬프트를 직접 작성하세요.
6. LLM이 만든 분석 아이디어가 실제 컬럼과 키 관계에 맞는지 검증하세요.

이 두 실습의 경우 과제 5번에서 진행 완료한 사항 

## 마무리

이번 장의 핵심은 “데이터를 불러왔다”에서 끝내지 않는 것입니다.

데이터 분석을 시작하기 전에 파일, 행과 열, 컬럼명, 데이터 타입, 결측치, 중복, 키 관계를 확인해야 이후 분석이 흔들리지 않습니다. 다음 장에서는 이 구조를 바탕으로 pandas의 선택, 필터링, 정렬, 집계 기초를 다룹니다.


## 6. Chapter 03 최종 판단 (끝)
### 데이터의 첫인상 3가지

1. 4개 csv 데이터 모두에서 주요 ID의 결측 및 중복이 0건으로 없었고, 현재 데이터에서 기본적인 식별에 대한 품질이 양호했다.
2. `orders.order_date` 경우 처음에는 문자열 타입이었으나 날짜 타입으로 변환 가능하였으며, 이를 통해 기간별 분석 전 날짜 컬럼 타입 및 변환 가능 여부를 확인해야 함을 알 수 있었다.
3. 키 사이 연결을 확인한 결과 csv 사이 기본적인 연결 관계가 정상적으로 유지되고 있는 데이터임을 알 수 있었다. 

### 다음 Chapter 전에 반드시 확인/처리해야 할 항목
1. 날짜 컬럼을 실제 분석에 사용할 수 있게 날짜 타입으로 변환하고, 분석 시에도 변환된 상태로 들어가는지 확인해야 한다.
2. 가입일과 주문일의 선후 관계를 확인해 가입일이 주문일보다 뒤에 이쓴 등의 정상적이지 않은 값이 있는지 확인해야 한다.
3. 매출을 분석하기 전에 각 status (ex. refunded, cancelled)를 분석에 어떻게 처리할 것인지 분석 기준을 정해야 한다. 

### 현재 데이터만으로 단정할 수 없는 것
현재 데이터에서 기본적인 구조는 확인했으나, 이것만으로 전체 데이터 품질이 좋다고 단정하기는 어렵다. 예를 들어 price와 unit_price의 차이가 어디에서 기인하는지, 배송비나 세금, 할인이 반영되는지는 현재 컬럼만으로 판단하기 어렵다. 분석 기준 역시 명확하지 않다. 범주형 데이터에서 빈도가 높거나 낮게 나타난 이유 역시 현재 데이터로는 알 수 없으므로, 실제 주문 특성인지 혹은 데이터 수집 조건 영향인지 추가적으로 확인해야 한다.

## 최종 제출 체크
- [X] Notebook을 처음부터 끝까지 실행했습니다.
- [X] 오류 셀이 남아 있지 않습니다.
- [X] 핵심 Evidence를 첨부했습니다.
- [X] 관찰과 해석을 구분했습니다.
- [X] 개인정보/Secret이 없습니다.
- [X] `chapter03/chapter03.ipynb`가 GitHub에서 정상 표시됩니다.
- [X] 최종 Notebook 파일 URL을 제출합니다.